File: everything/src/tools/get_plane_issue_id.js
import axios from 'axios';
import { match } from 'assert';

async function getPlaneIssueId(server, args) {
    try {
        // Parse parameters
        const params = args || {};

        // Validate parameters
        if (!params.issue_code) {
            const errorText = JSON.stringify({ error: "Error: Issue code is required" });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        const issue_code = params.issue_code;

        // Parse issue code format (e.g. "CLT-37")
        const match = issue_code.match(/^([A-Za-z]+)-(\d+)$/);
        if (!match) {
            const errorText = JSON.stringify({ error: "Error: Invalid issue code format. Expected format: PROJECT_CODE-NUMBER (e.g. CLT-37)" });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        const project_code = match[1];
        const sequence_id = parseInt(match[2], 10);

        // Configuration
        const API_KEY = process.env.PLANE_API_KEY || "plane_api_614f7240a5df4177840558c34bddb668";
        const BASE_URL = process.env.PLANE_BASE_URL || "http://192.168.50.90/api/v1";
        const WORKSPACE_SLUG = process.env.PLANE_WORKSPACE_SLUG || "test-space";

        const headers = {
            "X-API-Key": API_KEY,
            "Content-Type": "application/json"
        };

        // First get projects to find the one matching the code
        const projects_url = `${BASE_URL}/workspaces/${WORKSPACE_SLUG}/projects`;
        const projects_response = await axios.get(projects_url, { headers });

        if (projects_response.status !== 200) {
            const errorText = JSON.stringify({ error: `Error: Failed to get projects - ${projects_response.status}` });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        const projects_array = projects_response.data?.results; // Access the results array
        if (!Array.isArray(projects_array)) {
            const errorText = JSON.stringify({ error: "Error: Invalid response format when fetching projects." });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        let project = null;

        // Find project with matching identifier
        for (const p of projects_array) { // Iterate over the correct array
            if (p.identifier && p.identifier.toUpperCase() === project_code.toUpperCase()) {
                project = p;
                break;
            }
        }

        if (!project) {
            const errorText = JSON.stringify({ error: `Error: No project found with identifier ${project_code}` });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        // Get issues for the project
        const issues_url = `${BASE_URL}/workspaces/${WORKSPACE_SLUG}/projects/${project.id}/issues/`;
        const issues_response = await axios.get(issues_url, { headers });

        if (issues_response.status !== 200) {
            const errorText = JSON.stringify({ error: `Error: Failed to get issues - ${issues_response.status}` });
            return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        const issues_array = issues_response.data?.results; // Also apply the same logic for issues list
        if (!Array.isArray(issues_array)) {
             const errorText = JSON.stringify({ error: "Error: Invalid response format when fetching issues." });
             return { content: [{ type: 'text', text: errorText }], isError: true };
        }

        // Find issue with matching sequence_id
        for (const issue of issues_array) { // Iterate over the correct array
            if (issue.sequence_id === sequence_id) {
                const resultText = JSON.stringify({
                    issue_id: issue.id,
                    project_id: project.id,
                    name: issue.name,
                    current_state: issue.state
                }, null, 2);
                return { content: [{ type: 'text', text: resultText }] };
            }
        }

        const errorText = JSON.stringify({ error: `Error: No issue found with sequence ID ${sequence_id} in project ${project_code}` });
        return { content: [{ type: 'text', text: errorText }], isError: true };

    } catch (e) {
        if (axios.isAxiosError(e)) {
            const networkErrorText = JSON.stringify({ error: `Error: Network error - ${e.message}` });
            return { content: [{ type: 'text', text: networkErrorText }], isError: true };
        }
        const unexpectedErrorText = JSON.stringify({ error: `Error: Unexpected error - ${e.message}` });
        return { content: [{ type: 'text', text: unexpectedErrorText }], isError: true };
    }
}

export const getPlaneIssueIdToolDefinition = {
    name: 'get_plane_issue_id',
    description: 'Get the UUID of a Plane issue using its display code (e.g. \'CLT-37\').',
    inputSchema: {
        type: 'object',
        properties: {
            issue_code: {
                type: 'string',
                description: 'Issue code in the format PROJECT_CODE-NUMBER (e.g. "CLT-37")',
            },
        },
        required: ['issue_code'],
    },
};

export const get_plane_issue_id = {
  handler: getPlaneIssueId,
  definition: getPlaneIssueIdToolDefinition
}

File: everything/src/tools/get_plane_issue_id.py
import json
import requests
import os
import re

def get_plane_issue_id(params_json: str = "{}") -> str:
    """
    Get the UUID of a Plane issue using its display code (e.g. 'CLT-37').
    
    Args:
        params_json: A JSON string containing:
            {
                "issue_code": "PROJECT_CODE-NUMBER"  # e.g. "CLT-37"
            }
    
    Returns:
        str: Issue UUID if found, or error message
    
    Example:
        >>> get_plane_issue_id('{"issue_code": "CLT-37"}')
        "550e8400-e29b-41d4-a716-446655440000"
    """
    try:
        # Parse parameters
        try:
            params = json.loads(params_json) if params_json else {}
        except json.JSONDecodeError:
            return "Error: Invalid JSON parameters"
        
        # Validate parameters
        if not params.get("issue_code"):
            return "Error: Issue code is required"
        
        issue_code = params["issue_code"]
        
        # Parse issue code format (e.g. "CLT-37")
        match = re.match(r"([A-Za-z]+)-(\d+)", issue_code)
        if not match:
            return "Error: Invalid issue code format. Expected format: PROJECT_CODE-NUMBER (e.g. CLT-37)"
        
        project_code = match.group(1)
        sequence_id = int(match.group(2))
        
        # Configuration
        API_KEY = os.environ.get("PLANE_API_KEY", "plane_api_614f7240a5df4177840558c34bddb668")
        BASE_URL = os.environ.get("PLANE_BASE_URL", "http://192.168.50.90/api/v1")
        WORKSPACE_SLUG = os.environ.get("PLANE_WORKSPACE_SLUG", "test-space")
        
        headers = {
            "X-API-Key": API_KEY,
            "Content-Type": "application/json"
        }
        
        # First get projects to find the one matching the code
        projects_url = f"{BASE_URL}/workspaces/{WORKSPACE_SLUG}/projects"
        projects_response = requests.get(projects_url, headers=headers)
        
        if projects_response.status_code != 200:
            return f"Error: Failed to get projects - {projects_response.status_code}"
        
        projects = projects_response.json()
        project = None
        
        # Find project with matching identifier
        for p in projects:
            if p.get("identifier", "").upper() == project_code.upper():
                project = p
                break
        
        if not project:
            return f"Error: No project found with identifier {project_code}"
        
        # Get issues for the project
        issues_url = f"{BASE_URL}/workspaces/{WORKSPACE_SLUG}/projects/{project['id']}/issues/"
        issues_response = requests.get(issues_url, headers=headers)
        
        if issues_response.status_code != 200:
            return f"Error: Failed to get issues - {issues_response.status_code}"
        
        issues = issues_response.json()
        
        # Find issue with matching sequence_id
        for issue in issues:
            if issue.get("sequence_id") == sequence_id:
                return json.dumps({
                    "issue_id": issue["id"],
                    "project_id": project["id"],
                    "name": issue["name"],
                    "current_state": issue.get("state")
                })
        
        return f"Error: No issue found with sequence ID {sequence_id} in project {project_code}"
    
    except requests.RequestException as e:
        return f"Error: Network error - {str(e)}"
    except Exception as e:
        return f"Error: Unexpected error - {str(e)}"